examples/c/sockfilter.c

```diff
@@ -0,0 +1,177 @@
+// SPDX-License-Identifier: GPL-2.0 OR BSD-3-Clause
+/* Copyright (c) 2022 Jacky Yin */
+#include <arpa/inet.h>
+#include <assert.h>
+#include <bpf/libbpf.h>
+#include <linux/if_packet.h>
+#include <linux/if_ether.h>
+#include <linux/in.h>
+#include <net/if.h>
+#include <stdio.h>
+#include <sys/resource.h>
+#include <sys/socket.h>
+#include <unistd.h>
+#include "sockfilter.skel.h"
+
+struct so_event {
+	__be32 src_addr;
+	__be32 dst_addr;
+	union {
+		__be32 ports;
+		__be16 port16[2];
+	};
+	__u32 ip_proto;
+	__u32 pkt_type;
+};
+
+static const char * ipproto_mapping[IPPROTO_MAX] = {
+	[IPPROTO_IP] = "IP",
+	[IPPROTO_ICMP] = "ICMP",
+	[IPPROTO_IGMP] = "IGMP",
+	[IPPROTO_IPIP] = "IPIP",
+	[IPPROTO_TCP] = "TCP",
+	[IPPROTO_EGP] = "EGP",
+	[IPPROTO_PUP] = "PUP",
+	[IPPROTO_UDP] = "UDP",
+	[IPPROTO_IDP] = "IDP",
+	[IPPROTO_TP] = "TP",
+	[IPPROTO_DCCP] = "DCCP",
+	[IPPROTO_IPV6] = "IPV6",
+	[IPPROTO_RSVP] = "RSVP",
+	[IPPROTO_GRE] = "GRE",
+	[IPPROTO_ESP] = "ESP",
+	[IPPROTO_AH] = "AH",
+	[IPPROTO_MTP] = "MTP",
+	[IPPROTO_BEETPH] = "BEETPH",
+	[IPPROTO_ENCAP] = "ENCAP",
+	[IPPROTO_PIM] = "PIM",
+	[IPPROTO_COMP] = "COMP",
+	[IPPROTO_SCTP] = "SCTP",
+	[IPPROTO_UDPLITE] = "UDPLITE",
+	[IPPROTO_MPLS] = "MPLS",
+	[IPPROTO_RAW] = "RAW"
+};
+
+static inline int open_raw_sock(const char *name) {
+	struct sockaddr_ll sll;
+	int sock;
+
+	sock = socket(PF_PACKET, SOCK_RAW | SOCK_NONBLOCK | SOCK_CLOEXEC,
+				htons(ETH_P_ALL));
+	if (sock < 0) {
+		printf("cannot create raw socket\n");
```

Comment:
given it's an error, let's use fprintf(stderr) instead?

Reply:
OK, now I use fprintf(stderr) for all of the error messages.